File: articles/the-history-of-moral-psychology-from-moral-sense-theory-to-contemporary-research/c/diversification_score.c
/*
Synthetic diversification score for historical moral psychology emphases.
Educational scaffold only.

Compile:
  cc c/diversification_score.c -lm -o outputs/tables/diversification_score_c
Run:
  ./outputs/tables/diversification_score_c
*/

#include <math.h>
#include <stdio.h>

int main(void) {
    double values[] = {0.45, 0.40, 0.45, 0.75, 0.85, 0.80, 0.90};
    int n = 7;
    double total = 0.0;
    double entropy = 0.0;

    for (int i = 0; i < n; i++) {
        total += values[i];
    }

    for (int i = 0; i < n; i++) {
        double p = values[i] / total;
        if (p > 0.0) {
            entropy -= p * log(p);
        }
    }

    double normalized = entropy / log((double)n);

    printf("Synthetic normalized diversification score: %.3f\n", normalized);

    return 0;
}

File: articles/the-history-of-moral-psychology-from-moral-sense-theory-to-contemporary-research/rust/diversification_score.rs
// Synthetic diversification score for historical moral psychology emphases.
// Educational scaffold only.
//
// Compile:
//   rustc rust/diversification_score.rs -o outputs/tables/diversification_score_rust
// Run:
//   ./outputs/tables/diversification_score_rust

fn main() {
    let values = vec![0.45_f64, 0.40, 0.45, 0.75, 0.85, 0.80, 0.90];

    let total: f64 = values.iter().sum();

    let entropy: f64 = values
        .iter()
        .map(|value| value / total)
        .filter(|p| *p > 0.0)
        .map(|p| -p * p.ln())
        .sum();

    let normalized = entropy / (values.len() as f64).ln();

    println!("Synthetic normalized diversification score: {:.3}", normalized);
}
